modules/kinesis/src/it/scala/com/snowplowanalytics/snowplow/enrich/kinesis/EnrichKinesisSpec.scala

```diff
@@ -0,0 +1,173 @@
+
+          def consumeGood(ref: Ref[IO, AggregateGood]): Stream[IO, Unit] =
+            Source.init[IO](blocker, enrichedStreamConfig, ConfigCount.monitoring)
+              .map(KinesisRun.getPayload)
+              .evalMap(aggregateGood(_, ref))
+
+          def consumeBad(ref: Ref[IO, AggregateBad]): Stream[IO, Unit] =
+            Source.init[IO](blocker, badStreamConfig, ConfigCount.monitoring)
+              .map(KinesisRun.getPayload)
+              .evalMap(aggregateBad(_, ref))
+
+          def aggregateGood(r: Array[Byte], ref: Ref[IO, AggregateGood]): IO[Unit] =
+            for {
+              e <- IO(Event.parse(new String(r)).getOrElse(throw new RuntimeException("can't parse enriched event")))
+              _ <- ref.update(updateAggregateGood(_, e))
+            } yield ()
+
+          def aggregateBad(r: Array[Byte], ref: Ref[IO, AggregateBad]): IO[Unit] =
+            for {
+              br <- IO(new String(r))
+              _ <- ref.update(updateAggregateBad(_, br))
+            } yield ()
+
+          def updateAggregateGood(aggregate: AggregateGood, e: Event): AggregateGood =
+            e :: aggregate
+
+          def updateAggregateBad(aggregate: AggregateBad, br: String): AggregateBad =
+            br :: aggregate
+
+          for {
+            refGood <- Ref.of[IO, AggregateGood](Nil)
+            refBad <- Ref.of[IO, AggregateBad](Nil)
+            _ <-
+              consume(refGood, refBad)
+                .merge(generate)
+                .interruptAfter(3.minutes)
+                .attempt
+                .compile
+                .drain
+            aggregateGood <- refGood.get
+            aggregateBad <- refBad.get
+          } yield Aggregates(aggregateGood, aggregateBad)
+        }
+      }
+
+      for {
+        nbGood <- IO.pure(100l)
+        nbBad <- IO.pure(10l)
+        enrich <- IO(Containers.enrichCount)
+        aggregates <-
+          run(
+            nbGood,
+            nbBad,
+            ConfigCount.rawStreamConfig(Containers.localStackPort),
+            ConfigCount.enrichedStreamConfig(Containers.localStackPort),
+            ConfigCount.badStreamConfig(Containers.localStackPort)
+          )
+        _ <- IO(enrich.stop())
+        _ <- IO(println(s"Bad rows:\n${aggregates.bad.mkString("\n")}"))
+      } yield {
+        aggregates.good.size === nbGood
+        aggregates.bad.size === nbBad
+      }
+    }
+
+    "shutdown when it receives a SIGTERM" in {
+      for {
+        enrich <- IO(Containers.enrichStop)
+        _ <- IO.sleep(20.seconds) // give Enrich some time to initialize
+        isRunning = IO(enrich.isRunning())
+        _ <- IO(println("stop - Sending signal"))
+        _ <- IO(enrich.getDockerClient().killContainerCmd(enrich.getContainerId()).withSignal("TERM").exec())
+        _ <- isRunning.retryingOnFailures(
+          _ == false,
+          retryPolicy,
+          (_, _) => IO(())
+        )
+        logs = enrich.getLogs()
+        _ <- IO(enrich.stop())
```

Comment:
Do we need to call `stop` here? `TERM` signal should stop container without explicit call here, is that correct?

Reply:
Indeed :man_facepalming:

`enrich.stop()` has to be placed after the test `enrich.isRunning() === false`

It should be useless, it's just a security in case Enrich didn't stop after the SIGTERM

Reply:
It can't be placed after the test actually, because if the test fails it won't be reached :smile: I need to get `isRunning()` in a variable before `_ <- IO(enrich.stop())` and then use it in the test.

Reply:
Maybe we can utilize `Resource` from cats to wrap container start/stop?